File: nim/math/gcd_lcm.nim
# Print the divisor and multiple of fixed pairs, with Euclid's algorithm written as a while loop.
# Run: nim c -r gcd_lcm.nim

const Pairs = [(1071, 462), (270, 192), (17, 5), (120, 36)]

proc euclid(first, second: int): int =
  var
    a = first
    b = second
  while b != 0:
    let remainder = a mod b
    a = b
    b = remainder
  a

for (first, second) in Pairs:
  let divisor = euclid(first, second)
  echo first, " ", second, " ", divisor, " ", first div divisor * second

File: rust/math/gcd_lcm.rs
// Print the divisor and multiple of fixed pairs, with Euclid's algorithm run on a tuple swap.
// Build: rustc -o gcd_lcm gcd_lcm.rs

const PAIRS: [(u64, u64); 4] = [(1071, 462), (270, 192), (17, 5), (120, 36)];

fn gcd(mut first: u64, mut second: u64) -> u64 {
    while second != 0 {
        let remainder = first % second;
        first = second;
        second = remainder;
    }
    first
}

fn main() {
    for (first, second) in PAIRS {
        let divisor = gcd(first, second);
        println!("{} {} {} {}", first, second, divisor, first / divisor * second);
    }
}
